### PR TITLE
Add build spec

### PR DESCRIPTION
**Pull request by constantinpape**
Moves over the `build_spec` functionality from bioimage.io

TODO:
- [x] fix the test (need to create a pytest config)
- [ ] decide on naming for the module and functions; would probably be good to be more specific

Depends on #75 

cc @FynnBe @k-dominik

### CODE REVIEW

**Comment by constantinpape** [comment]
I fixed most tests, but I am not sure how to do the for `pytorch_state_dict`, which needs the `source`; and which must be a local file rn.

File: bioimageio/core/build_spec/build_spec.py [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+    from typing import get_args
+except ImportError:
+    from typing_extensions import get_args  # type: ignore
+
+#
+# utility functions to build the spec from python
+#
+
+
+def _get_local_path(uri, root=None):
```
Comment (FynnBe) [inline bioimageio/core/build_spec/build_spec.py:23]: this could actually be substituted with `bioimageio.spec.shared.utils.resolve_uri`

**Comment by constantinpape** [comment]
@FynnBe I am using `resolve_uri` now and fixed the issue in the test. So in principle this is good to go; but we could still think about more descriptive names for the `build_spec` submodule and its functions.

File: bioimageio/core/build_spec/build_spec.py [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+
+
+# TODO The citation entry should be improved so that we can properly derive doi vs. url
+def _build_cite(cite: Dict[str, str]):
+    citation_list = [model_spec.raw_nodes.CiteEntry(text=k, url=v) for k, v in cite.items()]
+    return citation_list
+
+
+# TODO we should make the name more specific: "build_model_spec"?
+# TODO maybe "build_raw_model" as it return raw_nodes.Model
```
Comment (FynnBe) [inline bioimageio/core/build_spec/build_spec.py:223]: I think 'build_raw_model' is a good idea, as we might add `build_model` in the future (for now that does not make so much sense as there is no serialization from non-raw nodes...

File: bioimageio/core/build_spec/build_spec.py [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+    new_weights = _get_weights(weight_uri, weight_type, None, root, **weight_kwargs)[0]
+    model.weights.update(new_weights)
+
+    serialized = model_spec.schema.Model().dump(model)
+    model = model_spec.schema.Model().load(serialized)
+
+    return model
+
+
+def serialize_spec(model, out_path):  # TODO change name to include model (see build_model_spec)
```
Comment (FynnBe) [inline bioimageio/core/build_spec/build_spec.py:436]: we can substitute this as well with `bioimageio.spec.save_raw_node`

**Review by FynnBe — APPROVED** [review]
Already LGTM. name changes could occur later as well...
maybe just substitute `serialize_spec` before merging.

File: bioimageio/core/build_spec/build_spec.py (hunk 3 above)
Comment (constantinpape) [reply bioimageio/core/build_spec/build_spec.py:436]: Ok, I will check if that works.

File: bioimageio/core/build_spec/build_spec.py (hunk 2 above)
Comment (constantinpape) [reply bioimageio/core/build_spec/build_spec.py]: I decided to go with `build_model` here. I find the `raw` things very confusing in the function names.

File: bioimageio/core/build_spec/build_spec.py (hunk 3 above)
Comment (constantinpape) [reply bioimageio/core/build_spec/build_spec.py]: removed

**Comment by constantinpape** [comment]
@FynnBe since I updated the biomageio.spec to the latest main the test imports don't work any more:
```
test_build_spec.py:4: in <module>
    from bioimageio.spec.model import schema
E   ModuleNotFoundError: No module named 'bioimageio'
```

Any idea what's causing this?

**Comment by constantinpape** [comment]
Looks like the import issue is just something wrong with my dev env on the cluster; works locally, so I'm merging this.

**Pull request merged**